File: 023_merge_k_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;

        PriorityQueue<ListNode> queue = new PriorityQueue<ListNode>((node1, node2) -> {
            return node1.val - node2.val;
        });

        for (ListNode node: lists) {
            // append the first node
            if (node != null) {
                queue.offer(node);
            }
        }

        ListNode head = new ListNode(0);
        ListNode curr = head;
        while (! queue.isEmpty()) {
            // get the node with smallest value
            // and link it
            curr.next = queue.poll();
            curr = curr.next;

            // push the remaining node at this linked list
            if (curr.next != null) {
                queue.offer(curr.next);
            }
        }
        return head.next;
    }
}

File: python/023_merge_k_sorted_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        import queue

        head = curr = ListNode(0)

        pq = queue.PriorityQueue()

        def add_node_in_pq(idx, node):
            if node:
                pq.put((node.val, idx, node))

        for idx, node in enumerate(lists):
            add_node_in_pq(idx, node)

        while not pq.empty():
            _, idx, node = pq.get()
            add_node_in_pq(idx, node.next)

            # remember to unlink the node

            node.next = None
            curr.next = node
            curr = curr.next

        return head.next

